asset-registry/src/lib.rs

```diff
@@ -0,0 +1,316 @@
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// Asset was not found.
+		AssetNotFound,
+		/// The version of the `VersionedMultiLocation` value used is not able
+		/// to be interpreted.
+		BadVersion,
+		/// The asset id is invalid.
+		InvalidAssetId,
+		/// Another asset was already register with this location.
+		ConflictingLocation,
+		/// Another asset was already register with this asset id.
+		ConflictingAssetId,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(crate) fn deposit_event)]
+	pub enum Event<T: Config> {
+		RegisteredAsset {
+			asset_id: T::AssetId,
+			metadata: AssetMetadata<T::Balance, T::CustomMetadata>,
+		},
+		UpdatedAsset {
+			asset_id: T::AssetId,
+			metadata: AssetMetadata<T::Balance, T::CustomMetadata>,
+		},
+		SetLocation {
+			asset_id: T::AssetId,
+			location: Box<VersionedMultiLocation>,
+		},
+	}
+
+	/// The metadata of an asset, indexed by asset id.
+	#[pallet::storage]
+	#[pallet::getter(fn metadata)]
+	pub type Metadata<T: Config> =
+		StorageMap<_, Twox64Concat, T::AssetId, AssetMetadata<T::Balance, T::CustomMetadata>, OptionQuery>;
+
+	/// Maps a multilocation to an asset id - useful when processing xcm
+	/// messages.
+	#[pallet::storage]
+	#[pallet::getter(fn location_to_asset_id)]
+	pub type LocationToAssetId<T: Config> = StorageMap<_, Twox64Concat, MultiLocation, T::AssetId, OptionQuery>;
+
+	/// The last processed asset id - used when assigning a sequential id.
+	#[pallet::storage]
+	#[pallet::getter(fn last_asset_id)]
+	pub(crate) type LastAssetId<T: Config> = StorageValue<_, T::AssetId, ValueQuery>;
+
+	#[pallet::genesis_config]
+	pub struct GenesisConfig<T: Config> {
+		_phantom: PhantomData<T>,
+	}
+
+	#[cfg(feature = "std")]
+	impl<T: Config> Default for GenesisConfig<T> {
+		fn default() -> Self {
+			Self {
+				_phantom: Default::default(),
+			}
+		}
+	}
+
+	#[pallet::genesis_build]
+	impl<T: Config> GenesisBuild<T> for GenesisConfig<T> {
+		fn build(&self) {}
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub(super) trait Store)]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(_);
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::weight(T::WeightInfo::register_asset())]
+		#[transactional]
+		pub fn register_asset(
+			origin: OriginFor<T>,
+			metadata: AssetMetadata<T::Balance, T::CustomMetadata>,
+			asset_id: Option<T::AssetId>,
+		) -> DispatchResult {
+			T::AuthorityOrigin::ensure_origin(origin)?;
+
+			Self::do_register_asset(metadata, asset_id)
+		}
+
+		#[allow(clippy::too_many_arguments)]
+		#[pallet::weight(T::WeightInfo::update_asset())]
+		#[transactional]
+		pub fn update_asset(
+			origin: OriginFor<T>,
+			asset_id: T::AssetId,
+			decimals: Option<u32>,
+			name: Option<Vec<u8>>,
+			symbol: Option<Vec<u8>>,
+			existential_deposit: Option<T::Balance>,
+			location: Option<Option<VersionedMultiLocation>>,
```

Comment:
we can simplify this parameter as `Option<VersionedMultiLocation>`.

Reply:
We can't - the outer Option determines whether or not to update that parameter, whereas the inner one is there because the user might want to unset the location, i.e. to make `metadata.location = None`